### ISSUE TITLE
can't build with golang-1.16

### ISSUE THREAD

**Issue by LocutusOfBorg** [libguestfs/libguestfs#65]
Hello, looks like new golang is requiring version to be defined?

```
Making all in golang
make[4]: Entering directory '/<<PKGBUILDDIR>>/debian/build-1/golang'
[ ../../../golang != . ] && ln -s /<<PKGBUILDDIR>>/debian/build-1/../../golang/src ./src
../run go install libguestfs.org/guestfs
run: warning: You used './configure --disable-appliance' so LIBGUESTFS_PATH
run: warning: has not been set automatically.
go install: version is required when current directory is not in a module
	Try 'go install libguestfs.org/guestfs@latest' to install the latest version
make[4]: *** [Makefile:2614: pkg/linux_amd64/libguestfs.org/guestfs.a] Error 1
```
https://blog.golang.org/go116-module-changes
this happens in Ubuntu, where go-1.16 is default now
https://launchpadlibrarian.net/532321596/buildlog_ubuntu-hirsute-amd64.libguestfs_1%3A1.44.1-1ubuntu1_BUILDING.txt.gz

**Comment by rwmjones** [comment]
We're going to need to do something like this: https://gitlab.com/nbdkit/libnbd/-/commit/78b09170fd5ca228f1f8b59f8643d0f5ad728473

**Comment by rwmjones** [comment]
This is fixed in https://github.com/libguestfs/libguestfs/commit/95de3f6c50feea52abf2794945e741674df620b1

The tests still fail with:
```
panic: runtime error: cgo argument has Go pointer to Go pointer
```

**Comment by LocutusOfBorg** [comment]
Thanks, I did a similar patch
```
--- libguestfs-1.44.1.orig/golang/Makefile.am
+++ libguestfs-1.44.1/golang/Makefile.am
@@ -21,6 +21,7 @@ include $(top_srcdir)/subdir-rules.mk
 pkg = libguestfs.org/guestfs
 
 source_files = \
+	src/$(pkg)/go.mod \
 	src/$(pkg)/guestfs.go \
 	src/$(pkg)/guestfs_*_test.go
 
@@ -37,17 +38,8 @@ EXTRA_DIST = \
 
 if HAVE_GOLANG
 
-golangpkgdir = $(GOROOT)/pkg/$(GOOS)_$(GOARCH)/$(pkg)
-golangsrcdir = $(GOROOT)/src/pkg/$(pkg)
-
-golangpkg_DATA = \
-	pkg/$(GOOS)_$(GOARCH)/$(pkg).a
-
-pkg/$(GOOS)_$(GOARCH)/$(pkg).a: src/$(pkg)/guestfs.go
-	-[ $(srcdir) != $(builddir) ] && ln -s $(abs_srcdir)/src $(builddir)/src
-	$(top_builddir)/run $(GOLANG) install $(pkg)
-
-golangsrc_DATA = $(source_files)
+all-local: $(source_files)
+	-[ $(srcdir) != $(builddir) ] && ln -s $(abs_srcdir)/src $(builddir)/src && cd src/$(pkg) && $(GOLANG) build
 
 TESTS_ENVIRONMENT = pkg=$(pkg) $(top_builddir)/run $(VG) # --test
 
--- /dev/null
+++ libguestfs-1.44.1/golang/src/libguestfs.org/guestfs/go.mod
@@ -0,0 +1,3 @@
+module main
+replace libguestfs.org/guestfs => ../../src/libguestfs.org/guestfs
+require libguestfs.org/guestfs v1.0.0
--- libguestfs-1.44.1.orig/m4/guestfs-golang.m4
+++ libguestfs-1.44.1/m4/guestfs-golang.m4
@@ -26,14 +26,6 @@ AS_IF([test "x$enable_golang" != "xno"],
         AC_MSG_CHECKING([if $GOLANG is usable])
         AS_IF([$GOLANG run $srcdir/golang/config-test.go 2>&AS_MESSAGE_LOG_FD],[
             AC_MSG_RESULT([yes])
-
-            # Substitute some golang environment.
-            GOOS=`$GOLANG env GOOS`
-            GOARCH=`$GOLANG env GOARCH`
-            GOROOT=`$GOLANG env GOROOT`
-            AC_SUBST([GOOS])
-            AC_SUBST([GOARCH])
-            AC_SUBST([GOROOT])
         ],[
             AC_MSG_RESULT([no])
             AC_MSG_WARN([golang ($GOLANG) is installed but not usable])
--- libguestfs-1.44.1.orig/run.in
+++ libguestfs-1.44.1/run.in
@@ -178,8 +178,6 @@ export LUA_CPATH
 
 # For golang.
 export GOLANG="@GOLANG@"
-prepend GOPATH "$b/golang"
-export GOPATH
 if [ -z "$CGO_CFLAGS" ]; then
     CGO_CFLAGS="-I$s/include"
 else
```

**Comment by LocutusOfBorg** [comment]
my build seems to go ok but fails due to an incomprensible failure
https://launchpadlibrarian.net/532489876/buildlog_ubuntu-hirsute-ppc64el.libguestfs_1%3A1.44.1-1ubuntu1ppa2_BUILDING.txt.gz
gcc: fatal error: Killed signal terminated program lto1


maybe I have to disable -flto in the new gcc...

**Comment by LocutusOfBorg** [comment]
mmm your patch fails in Ubuntu, probably because build dir and source dir are different!
```
touch stamp-guestfs-lua.pod
make[4]: Leaving directory '/<<PKGBUILDDIR>>/debian/build-1/lua/examples'
Making all in golang
make[4]: Entering directory '/<<PKGBUILDDIR>>/debian/build-1/golang'
cd src/libguestfs.org/guestfs && go build
/bin/bash: line 1: cd: src/libguestfs.org/guestfs: No such file or directory
make[4]: *** [Makefile:2556: all-local] Error 1
```

**Comment by LocutusOfBorg** [comment]
see e.g. https://launchpadlibrarian.net/532521057/buildlog_ubuntu-hirsute-ppc64el.libguestfs_1%3A1.44.1-1ubuntu1ppa4_BUILDING.txt.gz

**Comment by libguestfs** [comment]
On Thu, Apr 08, 2021 at 08:37:25AM -0700, Gianfranco Costamagna wrote:
> my build seems to go ok but fails due to an incomprensible failure
> https://launchpadlibrarian.net/532489876/
> buildlog_ubuntu-hirsute-ppc64el.libguestfs_1%3A1.44.1-1ubuntu1ppa2_BUILDING.txt.gz
> gcc: fatal error: Killed signal terminated program lto1
> 
> maybe I have to disable -flto in the new gcc...

I couldn't follow that link, but if GCC is crashing in lto1 then it
would probably be caused by LTO, so disabling it is the only way to
proceed.

We have also disabled LTO in Fedora builds of libguestfs, because
there's a bug in handling mixed C/OCaml files, see:

***@***.***/thread/Y3WTAS3MJ5WNNVVYPXARFXV6A6QH6KEY/

Rich.

-- 
Richard Jones, Virtualization Group, Red Hat http://people.redhat.com/~rjones
Read my programming and virtualization blog: http://rwmj.wordpress.com
virt-top is 'top' for virtual machines.  Tiny program with many
powerful monitoring features, net stats, disk stats, logging, etc.
http://people.redhat.com/~rjones/virt-top

**Comment by rwmjones** [comment]
Does this work?  It's based off your original patch: https://github.com/libguestfs/libguestfs/commit/c8cab6219ff9e0feeb31a9fd18d3ff8f2a9c5de5

**Comment by LocutusOfBorg** [comment]
```
https://launchpadlibrarian.net/532669514/buildlog_ubuntu-hirsute-ppc64el.libguestfs_1%3A1.44.1-1ubuntu3_BUILDING.txt.gz
make[4]: Leaving directory '/<<PKGBUILDDIR>>/debian/build-1/lua'
Making all in lua/examples
make[4]: Entering directory '/<<PKGBUILDDIR>>/debian/build-1/lua/examples'
perl /<<PKGBUILDDIR>>/debian/build-1/podwrapper.pl --path ../../../../lua/examples \
  --section 3 \
  --man guestfs-lua.3 \
  --html ../../website/guestfs-lua.3.html \
  --verbatim ../../../../lua/examples/create_disk.lua:@EXAMPLE1@ \
  --verbatim ../../../../lua/examples/inspect_vm.lua:@EXAMPLE2@ \
  --license examples \
  ../../../../lua/examples/guestfs-lua.pod
touch stamp-guestfs-lua.pod
make[4]: Leaving directory '/<<PKGBUILDDIR>>/debian/build-1/lua/examples'
Making all in golang
make[4]: Entering directory '/<<PKGBUILDDIR>>/debian/build-1/golang'
[ ../../../golang != . ] && ln -s /<<PKGBUILDDIR>>/debian/build-1/../../golang/src ./src
cd src/libguestfs.org/guestfs && go build
# libguestfs.org/guestfs
./guestfs.go:31:10: fatal error: guestfs.h: No such file or directory
   31 | #include "guestfs.h"
      |          ^~~~~~~~~~~
compilation terminated.
make[4]: *** [Makefile:2557: all-local] Error 2
make[4]: Leaving directory '/<<PKGBUILDDIR>>/debian/build-1/golang'
```

**Comment by rwmjones** [comment]
This should fix it probably: afe7e16c0fa1329751f12f700331b386eb21d6a4

**Comment by LocutusOfBorg** [comment]
yes it worked thanks!
